marlowe-dashboard-client/src/Contract/State.purs

```diff
@@ -60,12 +58,184 @@ import Web.HTML.HTMLElement as HTMLElement
+    }
+
+mkInitialState :: Slot -> ContractInstanceId -> History -> Maybe State
+mkInitialState currentSlot contractInstanceId history =
+  let
+    marloweParams = get1 $ unwrap history
+
+    marloweData = get2 $ unwrap history
+
+    transactionInputs = get3 $ unwrap history
+
+    contract = marloweData.marloweContract
+
+    mTemplate = findTemplate contract
+
+    -- We can't use the currentSlot to create the initial execution state, since the contract might
+    -- have been created several slots ago. Hopefully this doesn't matter (the argument is only used
+    -- to set the minSlot in the contract's initial state).
+    initialExecutionState = initExecution zero contract
+  in
+    flip map mTemplate \template ->
+      let
+        initialState =
+          { tab: Tasks
+          , executionState: initialExecutionState
+          , previousSteps: mempty
+          , marloweParams
+          , contractInstanceId
+          , selectedStep: 0
+          , metadata: template.metaData
+          , participants: mempty --- FIXME: I don't think there is any way we can do this
+          , mActiveUserParty: Nothing -- FIXME: this should be a function of the walletDetails
+          , namedActions: mempty
+          }
+
+        updateExecutionState = over _executionState (applyTransactionInputs transactionInputs)
+      in
+        initialState
+          # updateExecutionState
+          # regenerateStepCards currentSlot
+          # selectLastStep
+
+updateState :: Slot -> History -> State -> State
+updateState currentSlot history state =
+  let
+    allTransactionInputs = get3 $ unwrap history
+
+    previousTransactionInputs = map (\previousState -> previousState.txInput) $ view (_executionState <<< _previousState) state
+
+    newTransactionInputs = difference allTransactionInputs previousTransactionInputs
+
+    updateExecutionState = over _executionState (applyTransactionInputs newTransactionInputs)
+  in
+    state
+      # updateExecutionState
+      # regenerateStepCards currentSlot
+      # selectLastStep
+
+handleAction ::
+  forall m.
+  MonadAff m =>
+  MonadAsk Env m =>
+  ManageMarlowe m =>
+  Toast m =>
+  WalletDetails -> Action -> HalogenM State Action ChildSlots Msg m Unit
+handleAction walletDetails (ConfirmAction namedAction) = do
+  currentExeState <- use _executionState
+  marloweParams <- use _marloweParams
+  slot <- liftEffect currentSlot
+  let
+    input = toInput namedAction
+
+    txInput = mkTx slot (currentExeState ^. _currentContract) (Unfoldable.fromMaybe input)
+  -- TODO: currently we just ignore errors but we probably want to do something better in the future
+  ajaxApplyInputs <- marloweApplyTransactionInput walletDetails marloweParams txInput
+  case ajaxApplyInputs of
+    Left ajaxError -> addToast $ ajaxErrorToast "Failed to submit transaction." ajaxError
+    Right _ -> do
+      -- FIXME: don't applyTx here, but let the update from the websocket tell us about the change
```

Comment:
We should probably add a ticket and implement the "loading" status. Looking at Russ designs we can even add the [tailwind's pulse](https://tailwindcss.com/docs/animation#pulse) animation to it :)

Reply:
:+1: